File: Recursion/Day2/BinarySearch.java

//*** Binary search recursively ***
//Binary search can only be applied on the array if the array is sorted

import java.util.*;

class BinarySearch {
    
    int binarySearch(int arr[], int start,int end, int val) {

        if(end>=start) {

            int mid = start+(end-start)/2;

            if(arr[mid]==val)
                return mid;
            else if(arr[mid] > val)
                return binarySearch(arr, start, mid-1, val);
            return binarySearch(arr, mid+1, end, val);
        }
        return -1;
    }
}

class BS_Main {

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        BinarySearch obj = new BinarySearch();

        System.out.print("Enter the size of the array: ");
        int size = sc.nextInt();

        int arr[] = new int[size];

        System.out.print("Enter the elements in the array: ");
        for(int i=0;i<size;i++) {
            arr[i] = sc.nextInt();
        }

        System.out.print("Enter the element you want to search: ");
        int element = sc.nextInt();

        int ind = obj.binarySearch(arr,0,size-1,element);

        if(ind==-1) 
            System.out.println("The element is not present in the array!!!");
        else
            System.out.println("The element is present at index: "+ind);
    }
}

File: DataStructure/MyPractice/DataStructures/Searching/BinarySearch.cpp

#include <iostream>
using namespace std;


int binarySearch(int arr[],int i,int j,int val) {

	int mid;
      	mid = (i+j)/2;
	
	if(j>=i) {
		if(arr[mid] == val)
			return mid;
		else if(arr[mid] > val)
			return binarySearch(arr,i,mid-1,val);
		else if(arr[mid] < val)
			return binarySearch(arr,mid+1,j,val);
	}
	return 0;
}


int main() {

	int arr[] = {1,4,5,6,10,34,343};
	int x= 34;

	int val = binarySearch(arr,0,sizeof(arr)/sizeof(int),x);
	if(val!=0)
		cout << "Value is present at index: " << val << endl;
	else 
		cout << "Value is not present in the array " << endl;
	
		
	return 0;
}
